server/conn.go

```diff
@@ -738,22 +745,27 @@ func (cc *clientConn) handleLoadData(goCtx goctx.Context, loadDataInfo *executor
 // As the execution time of this function represents the performance of TiDB, we do time log and metrics here.
 // There is a special query `load data` that does not return result, which is handled differently.
 func (cc *clientConn) handleQuery(goCtx goctx.Context, sql string) (err error) {
-	rs, err := cc.ctx.Execute(goCtx, sql)
+	goCtx1, cancelFunc := goctx.WithCancel(goCtx)
```

Comment:
You also need to update cc.handleStmtPrepare and cc.handleStmtExecute.

Reply:
How about set cancelFunc in `dispatch`?